### DISCUSSION TITLE
iOS - Using Threads

### DISCUSSION THREAD

**Discussion by always2code** [MobileNativeFoundation/discussions#134] [General]
I never did intensive implementation around Threads, in my experience I only used `DispatchQueue.global(qos: .background).async` &  `DispatchQueue.main.asyncAfter`. 

I am curious if anyone had to do extensive implementations around threads? if yes, what was the original problem or scenario in project and what approach you used.  

(Asking this out of curiosity, highly appreciate your inputs, it really helps for my learning)

**Comment by Mazyod** [comment]
Hello,

Although I am an iOS developer by trade, I spend time building other stuff, and especially fond of writing highly parallel systems with Elixir. The latter influenced the former in how I started looking at iOS applications when it comes to concurrency/parallelism.

Although Grand Central Dispatch (GCD) APIs are quite helpful, they are indeed considered low-level when it comes to building highly concurrent systems. Something more user-friendly and high level is the Actor Model, which is also introduced to Swift recently. I would highly recommend you start there. It does abstract the details of thread handling and impose restrictions on how your code will look like, but it comes with great benefits as well.

In case you're looking for another approach that is perhaps more battle-tested in iOS land, you could start with RxSwift and the constructs they provide to easily build data "flows" that push data to background threads, then back to the main thread, in a fairly robust and and concise matter.

All-in-all, my 2¢ on the topic is that parallel programming can get complicated and frustrating if the code is designed with an imperative mindset and OOP paradigm. It is much more simpler and manageable once you adopt the principles of functional programming more and break down the code into asynchronous computational chunks.

**Comment by Mazyod** [comment]
As for examples:

1. Networking layer
A networking layer usually involves the following (more or less)
- Building the Request
- Dispatching the Request / reading the response
- Parsing the response
.. With possibly many different other operations in the middle, like:
- Logging and Error handling
- SSL pinning
- Encryption / Decryption
- ...

To build a modular Networking layer, we made each step into a simple functional component, and as a bonus, it allowed us to parallelize our networking layer thanks to this modular approach. We connect the "phases" in an Rx pipeline, switching between threads easily as required.

2. Download Queues
For another app, we had a simple requirement for managing possibly 10s of concurrent downloads. This was some time back, in the Objective C days, and the popular approach was using OperationQueue class to manage all the concurrent work using NSOperation objects. You basically subclass NSOperation, write your runloop within, and then instantiate them as needed, and they will be managed for you by the OperationQueue.

**Comment by Koshub** [comment]
I agree with @Mazyod at some points. I would recommend trying to treat the "threading" problems as "details" so you isolate them from the "business"/other logic. That will help to:
- reduce the threading management complexity, especially in cases where you need these 
  > extensive implementations around threads
  
  by any reason. Since it will be isolated you can tune/improve it later without changing too much.
- make an ability to distribute work across teams. Say, the core team works on performance (with threading problems), other teams work on features and do not care about threads
- write simple and fast unit tests without async API
- adopt easy new Apple features (like moving from DispatchQueue block API to async/await, etc)
